src/serializer/ResidualHeapVisitor.js

```diff
@@ -1210,15 +1210,15 @@ export class ResidualHeapVisitor {
         residualBinding.hasLeaked = true;
         // This may not have been referentialized if the binding is a local of an optimized function.
         // in that case, we need to figure out which optimized function it is, and referentialize it in that scope.
-        let optimizedFunctionScope = this._getAdditionalFunctionOfScope();
+        let commonScope = this._getCommonScope();
         if (residualBinding.potentialReferentializationScopes.size === 0) {
```

Comment:
This `if` statement was already there before your change, but I am not sure it's correct. What if you have an object that gets assigned to in both an optimized function and from a pure scope in global code?

Reply:
I don’t know enough about this to know whether or not it is correct. I tried producing a failing test case based on your comment but didn’t have success. I’ll leave to @cblappert who has better context on this code and this condition.

The defense for this condition based on my understanding of the code is that we only need to execute the code under this condition once per `binding` in the “largest” possible scope. Perhaps accidentally if this is in a pure scope it will be visited first since optimized functions are only evaluated at the very end of global scope? Under my understanding, if `__optimize` evaluated the function in place (instead of at the end of global scope) the following example would referentialize `i` in `f`.

```js
const havoc = __abstract("function", "(() => {})");

global.result = __evaluatePureFunction(() => {
  const f = () => i++;
  __optimize(f);

  let i = 0;

  havoc(f);

  return f;
});
```

Although I might completely misunderstand 🙂